### PR TITLE
HPCC-22312 Failing Std.Date unit tests on macOS

### PR DESCRIPTION
**Pull request by dcamper**
This change specifically involves the %Y formatting directive for year output (not parsing).

Complete list of Date functions that were affected:

DateToString
SecondsToString
TimestampToString
ConvertDateFormat
ConvertDateFormatMultiple

Also included with this change is a new __OS__ ECL constant value:  'linux-macos'.

Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Manual testing plus exercising (new and modified) tests in TestStd.Date.TestFormat.

### CODE REVIEW

**Comment by dcamper** [comment]
@richardkchapman @ghalliday Please review.

File: ecl/hql/hqllex.l [hunk 1]
```diff
@@ -1044,6 +1044,8 @@ __OS__              {
                             return lookupIdentifierToken(returnToken, lexer, lookup, activeState, CUR_TOKEN_TEXT);
 #ifdef _WIN32
                         returnToken.setExpr(createConstant("windows"));
+#elif defined(__APPLE__)
+                        returnToken.setExpr(createConstant("linux-macos"));
```
Comment (ghalliday) [inline ecl/hql/hqllex.l]: @richardkchapman is this a sensible value?

File: ecllibrary/std/Date.ecl [hunk 2]
```diff
@@ -56,6 +56,47 @@ END;
...
+ * precision format where needed.
+ *
+ * As of this writing, here is how our supported operating systems natively
+ * deal with small years (e.g. 11) and the %Y directive by default:
+ *
+ * macOS:	0011
+ * Windows:	0011
+ * Linux:	11
+ *
+ * All support Linux distributions have glibc, which provides support for
```
Comment (ghalliday) [inline ecllibrary/std/Date.ecl]: What happens if the system is compiled using clang?

**Review by ghalliday — COMMENTED** [review]
@richardkchapman added a couple of questions

File: ecllibrary/std/Date.ecl (hunk 2 above)
Comment (dcamper) [reply ecllibrary/std/Date.ecl]: Good question.  I installed Clang 7.0 on my Ubuntu 18.04 system and used (basically) the same code I used on Windows for testing (which is just spitting out the formatted year value from a valid struct tm).  The result was identical to gcc's result:  Using %Y, small years were returned without leading zeroes, and the precision specifier was supported (so %04Y did pad the result).

TL;DR version:  Clang appears to work the same as gcc, which makes sense since it tries to emulate the Gnu toolchain.  Apple may have modified their version of LLVM/Clang so that the leading zeroes are included.

File: ecl/hql/hqllex.l (hunk 1 above)
Comment (richardkchapman) [reply ecl/hql/hqllex.l]: Not really. MacOS is not linux or linux-based, though it shares some similarity. I don't see how returning this value helps any with back-compatibility - existing code is still going to have to change - so we may as well get it right.

**Comment by richardkchapman** [comment]
Has this been tried on Windows?

I'm not at all convinced that this is moving in the right direction, nor that it won't cause major compatibility headaches for existing code (99.99% of it running on Linux) that was relying on the previous behaviour. I suspect we should be changing MacOS implementation to return 2 digit dates (or marking as wontfix, if that's too hard)

**Comment by dcamper** [comment]
It has been tested on Windows.

Richard, I agree that this feels like a heavy-handed fix.  It does normalize the returned year value, but what changes affects the majority of our actual installations and it is bit hackish on top of that.  It does not feel worth it.

However, rather than just marking the issue as Won't Fix leaves the Std.Date ECL unit tests failing on macOS.

What do you think about changing this PR so that:

1. We add a (more reasonable) __OS__ value for macOS
2. Use __OS__ to make OS-level unit tests for those that will fail due to this year inconsistency, so they will always pass on all supported platforms
3. Perhaps document the inconsistency, if it is deemed important enough

Thoughts?

**Comment by dcamper** [comment]
Note:  If this PR is accepted then documentation for __OS__ in the language reference manual will need to be updated.

**Comment by richardkchapman** [comment]
There is also thorlib.os() that should perhaps be changed? Can be a separate Jira though
